source/adios2/operator/compress/CompressPNG.cpp

```diff
@@ -0,0 +1,220 @@
+ * accompanying file Copyright.txt for details.
+ *
+ * CompressPNG.cpp
+ *
+ *  Created on: Jun 10, 2019
+ *      Author: William F Godoy godoywf@ornl.gov
+ */
+#include "CompressPNG.h"
+
+#include <cstring> // std::memset
+
+extern "C" {
+#include <png.h>
+}
+
+#include "adios2/helper/adiosFunctions.h"
+
+namespace adios2
+{
+namespace core
+{
+namespace compress
+{
+
+const std::map<std::string, uint32_t> CompressPNG::m_ColorTypes = {
+    {"PNG_COLOR_TYPE_GRAY", PNG_COLOR_TYPE_GRAY},
+    {"PNG_COLOR_TYPE_PALETTE", PNG_COLOR_TYPE_PALETTE},
+    {"PNG_COLOR_TYPE_RGB", PNG_COLOR_TYPE_RGB},
+    {"PNG_COLOR_TYPE_RGB_ALPHA", PNG_COLOR_TYPE_RGB_ALPHA},
+    {"PNG_COLOR_TYPE_GRAY_ALPHA", PNG_COLOR_TYPE_GRAY_ALPHA},
+    {"PNG_COLOR_TYPE_RGBA", PNG_COLOR_TYPE_RGBA},
+    {"PNG_COLOR_TYPE_GA", PNG_COLOR_TYPE_GA}};
+
+const std::map<std::string, std::set<uint32_t>> CompressPNG::m_BitDepths = {
+    {"PNG_COLOR_TYPE_GRAY", {1, 2, 4, 8, 16}},
+    {"PNG_COLOR_TYPE_PALETTE", {1, 2, 4, 8}},
+    {"PNG_COLOR_TYPE_RGB", {8, 16}},
+    {"PNG_COLOR_TYPE_RGB_ALPHA", {8, 16}},
+    {"PNG_COLOR_TYPE_GRAY_ALPHA", {8, 16}},
+    {"PNG_COLOR_TYPE_RGBA", {8, 16}},
+    {"PNG_COLOR_TYPE_GA", {8, 16}}};
+
+// PUBLIC
+CompressPNG::CompressPNG(const Params &parameters, const bool debugMode)
+: Operator("png", parameters, debugMode)
+{
+}
+
+size_t CompressPNG::Compress(const void *dataIn, const Dims &dimensions,
+                             const size_t elementSize,
+                             const std::string /*type*/, void *bufferOut,
+                             const Params &parameters, Params &info) const
+{
+    auto lf_Write = [](png_structp png_ptr, png_bytep data, png_size_t length) {
+        DestInfo *pDestInfo =
+            reinterpret_cast<DestInfo *>(png_get_io_ptr(png_ptr));
+        std::memcpy(pDestInfo->BufferOut + pDestInfo->Offset, data, length);
+        pDestInfo->Offset += length;
+    };
+
+    const std::size_t ndims = dimensions.size();
+    if (m_DebugMode)
+    {
+        if (ndims != 3 && ndims != 2)
+        {
+            throw std::invalid_argument(
+                "ERROR: image number of dimensions " + std::to_string(ndims) +
+                " is invalid, must be 2 {height,width*bytes_per_pixel} or 3"
+                " {height,width,bytes_per_pixel]} , in call to ADIOS2 PNG "
+                " compression\n");
+        }
+    }
+
+    // defaults
+    int compressionLevel = 1;
+    int colorType = PNG_COLOR_TYPE_RGBA;
+    int bitDepth = 8;
+    std::string colorTypeStr = "PNG_COLOR_TYPE_RGBA";
+
+    for (const auto &itParameter : parameters)
+    {
+        const std::string key = itParameter.first;
+        const std::string value = itParameter.second;
+
+        if (key == "compression_level")
+        {
+            compressionLevel = static_cast<int>(helper::StringTo<int32_t>(
+                value, m_DebugMode, "when setting PNG level parameter\n"));
+            if (m_DebugMode)
```

Comment:
So I still am a bit fuzzy on the philosophy of this library, but I feel like this error message should be given independently of whether `m_DebugMode` is true or false. (You'll have to give me a rundown on the vision for `m_DebugMode` in meatspace.)

Reply:
> (You'll have to give me a rundown on the vision for m_DebugMode in meatspace.)

Sure. DebugMode is for workflow developers, it catches `std::logic_error` (including `std::invalid_argument`) provided by the user. Turning it off is just convenient for static workflows and prevents user input checks already checked during development. On the other hand, `std::runtime_error` is always checked.